### DISCUSSION TITLE
Command Line Tools

### DISCUSSION THREAD

**Discussion by rogerxu** [rogerxu/rogerxu.github.io#286] [Show and tell]
[Rust Easy! Modern Cross-platform Command Line Tools to Supercharge Your Terminal | Technorage (deepu.tech)](https://deepu.tech/rust-terminal-tools-linux-mac-windows-fish-zsh/)

**Comment by rogerxu** [comment]
## Directory

### ls

#### exa

#### lsd

#### broot

[How to Install and Use the broot Command | Linode Docs](https://www.linode.com/docs/guides/installing-and-using-the-broot-command/)

### cd

#### zoxide

[How to Install and Use zoxide on Linux | Linode Docs](https://www.linode.com/docs/guides/how-to-use-zoxide/)

**Comment by rogerxu** [comment]
## File

### cat

[The Linux cat Command | Linode Docs](https://www.linode.com/docs/guides/linux-cat-command/)

#### bat

[How to Install and Use the Linux bat Command | Linode Docs](https://www.linode.com/docs/guides/how-to-install-and-use-the-bat-command-on-linux/)

```sh
apk add bat
```

### less

### head

head

[View the Beginning of Text Files with head | Linode Docs](https://www.linode.com/docs/guides/view-the-beginning-of-text-files-with-head/)

tail

[View and Follow the End of Text Files with tail | Linode Docs](https://www.linode.com/docs/guides/view-and-follow-the-end-of-text-files-with-tail/)

### ln

[Create File System Links with ln | Linode Docs](https://www.linode.com/docs/guides/create-file-system-links-with-ln/)

### tar

[Archiving and Compressing files with GNU Tar and GNU Zip | Linode Docs](https://www.linode.com/docs/guides/archiving-and-compressing-files-with-gnu-tar-and-gnu-zip/)

### cp

#### rsync

[Introduction to rsync | Linode Docs](https://www.linode.com/docs/guides/introduction-to-rsync/)

[Using the Rsync Cross Platform Command-Line Utility | Linode Docs](https://www.linode.com/docs/guides/rsync-cross-platform-tutorial/)

```bash
$ rsync -avhPud source/ root@hostname.lan:destination
```

#### rclone

[How to Use Rclone to Sync Files to Linode Object Storage | Linode Docs](https://www.linode.com/docs/guides/rclone-object-storage-file-sync/)

**Comment by rogerxu** [comment]
## Disk

### df

#### duf

[How to Install and Use the duf Command on Linux | Linode Docs](https://www.linode.com/docs/guides/duf-command-on-linux-installation/)

### du

#### dust

[Installing and Using dust on Linux | Linode Docs](https://www.linode.com/docs/guides/dust-command-on-linux-installation/)

```sh
apk add dust
```

**Comment by rogerxu** [comment]
## Network

### Ping

[Introduction to the Linux ping Command | Linode Docs](https://www.linode.com/docs/guides/linux-ping-command/)

#### gping

[How to Install and Use gping on Linux | Linode Docs](https://www.linode.com/docs/guides/how-to-use-gping-on-linux/)

### curl

[How to Use cURL with RESTful APIs | Linode Docs](https://www.linode.com/docs/guides/curl-for-rest-api/)

#### wget

[Download Resources from the Command Line with wget | Linode Docs](https://www.linode.com/docs/guides/download-resources-from-the-command-line-with-wget/)

#### HTTPie 

[How to Install and Use HTTPie on Linux | Linode Docs](https://www.linode.com/docs/guides/installing-and-using-httpie-on-linux/)

#### curlie

[How to Install and Use the curlie Command on Linux | Linode Docs](https://www.linode.com/docs/guides/installing-and-using-the-curlie-command-on-linux/)

### DNS

```sh
apk add bind-tools
```

```bash
apt install dnsutils
```

#### nslookup

[A Complete Guide to the nslookup Command | Linode Docs](https://www.linode.com/docs/guides/how-to-use-nslookup-command/)

```bash
$ nslookup whoami.akamai.net 8.8.8.8
Server:         8.8.8.8
Address:        8.8.8.8#53

Non-authoritative answer:
Name:   whoami.akamai.net
Address: 112.65.95.5

```

```bash
$ nslookup -type=any whoami.akamai.net 8.8.8.8
Server:         8.8.8.8
Address:        8.8.8.8#53

Non-authoritative answer:
Name:   whoami.akamai.net
Address: 172.253.6.4

Authoritative answers can be found from:

```

```bash
$ nslookup -debug whoami.akamai.net 8.8.8.8
Server:         8.8.8.8
Address:        8.8.8.8#53

------------
    QUESTIONS:
        whoami.akamai.net, type = A, class = IN
    ANSWERS:
    ->  whoami.akamai.net
        internet address = 112.65.95.4
        ttl = 11
    AUTHORITY RECORDS:
    ADDITIONAL RECORDS:
------------
Non-authoritative answer:
Name:   whoami.akamai.net
Address: 112.65.95.4
------------
    QUESTIONS:
        whoami.akamai.net, type = AAAA, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ADDITIONAL RECORDS:
------------

```

#### dig

```bash
$ dig -v
DiG 9.18.24
```

```bash
$ dig q-type q-class {global-d-opt} host @local-server {local-d-opt}
```

```bash
$ dig +tcp +short whoami.akamai.net @8.8.8.8
112.65.95.5
```

#### dog

[ogham/dog: A command-line DNS client. (github.com)](https://github.com/ogham/dog)

[Use the Linux dog Command to Look Up DNS Records | Linode Docs](https://www.linode.com/docs/guides/use-dog-linux-dns-client/)

```sh
apk add dog
```

```bash
$ dog whoami.akamai.net @8.8.8.8
```

**Comment by rogerxu** [comment]
## Search

### find

[How to Find Files in Linux Using the Command Line | Linode Docs](https://www.linode.com/docs/guides/find-files-in-linux-using-the-command-line/)

#### fd

[How to Find Files With the fd Command | Linode Docs](https://www.linode.com/docs/guides/finding-files-with-fd-command/)

```sh
apk add fd
```

### grep

[How to Use the Grep Command to Find Information in Files | Linode Docs](https://www.linode.com/docs/guides/how-to-use-grep-command/)

#### ripgrep

[How to Install and Use ripgrep | Linode Docs](https://www.linode.com/docs/guides/ripgrep-linux-installation/)

```sh
apk add ripgrep
```

#### fzf

[How to Install and Use fzf on Linux | Linode Docs](https://www.linode.com/docs/guides/how-to-use-fzf/)

```sh
apk add fzf
```

### cut

#### choose

[Use the choose Command for Text Processing on Linux | Linode Docs](https://www.linode.com/docs/guides/use-linux-choose-command/)

```sh
apk add choose
```

```zsh
brew install choose-rust
```

### sed

[Manipulate Text from the Command Line with sed | Linode Docs](https://www.linode.com/docs/guides/manipulate-text-from-the-command-line-with-sed/)

#### sd

[How to Use the sd Command for Find and Replace on Linux | Linode Docs](https://www.linode.com/docs/guides/linux-sd-command/)

```sh
apk add sd
```

### whatis

### tldr

[tldr pages](https://tldr.sh/)

[tldr InBrowser.App](https://tldr.inbrowser.app/)

[tldr pages clients · tldr-pages/tldr Wiki (github.com)](https://github.com/tldr-pages/tldr/wiki/tldr-pages-clients)

[How to Use tldr pages on Linux | Linode Docs](https://www.linode.com/docs/guides/tldr-pages-on-linux/)

#### tlrc

[tldr-pages/tlrc: A tldr client written in Rust (github.com)](https://github.com/tldr-pages/tlrc)

```zsh
brew install tlrc
```

#### tealdeer

[dbrgn/tealdeer: A very fast implementation of tldr in Rust. (github.com)](https://github.com/dbrgn/tealdeer)

```zsh
pacman -Syu tealdeer
```

```bash
apt install tealdeer
```

```zsh
brew install tealdeer
```

### cheat

[How to Install and Use the cheat Command on Linux | Linode Docs](https://www.linode.com/docs/guides/linux-cheat-command/)

**Comment by rogerxu** [comment]
## System

### ps

#### procs

[View Active Processes in Linux with the procs Command | Linode Docs](https://www.linode.com/docs/guides/view-active-linux-processes-with-procs/)

```sh
apk add procs
```

### top

#### btop

[aristocratos/btop: A monitor of resources](https://github.com/aristocratos/btop)

```sh
apk add btop
```

#### bottom

[How to Install and Use bottom on Linux | Linode Docs](https://www.linode.com/docs/guides/installing-and-using-bottom-on-linux/)

```sh
apk add bottom
```

#### gtop

[How to Install and Use gtop on Linux | Linode Docs](https://www.linode.com/docs/guides/installing-and-using-gtop-on-linux/)

#### htop

#### glances

[How to Use Glances for System Monitoring on Linux | Linode Docs](https://www.linode.com/docs/guides/how-to-use-glances-system-monitoring/)

```sh
apk add glances
```

### history

#### mcfly

[Installing and Using McFly to Search Bash or ZSH History | Linode Docs](https://www.linode.com/docs/guides/using-mcfly-to-search-bash-or-zsh-history/)

```sh
apk add mcfly
```

### env

[How to Set and Use Linux Environment Variables | Linode Docs](https://www.linode.com/docs/guides/how-to-set-linux-environment-variables/)

### date

[How to Use the Date Command in Linux | Linode Docs](https://www.linode.com/docs/guides/use-the-date-command-in-linux/)

### lsof

**Comment by rogerxu** [comment]
## JSON

### jq

```zsh
jq '.payload | fromjson | .[] | select(.token | startswith("clash")) | pick(.token, .sha256)' cask.jws.json
```

```json
{
  "token": "clash-verge-rev",
  "sha256": "8268db2a3c45f910aa9092e3ac5a44876cbf139398bc5604ab362f35c764f9ab"
}
```

**Comment by rogerxu** [comment]
## Runtime

### C++

#### gcc

TBD

### Python

TBD

### Java

TBD

### Node.js

#### fnm

[Schniz/fnm: 🚀 Fast and simple Node.js version manager, built in Rust (github.com)](https://github.com/Schniz/fnm)

```zsh
brew install fnm
```

### Deno

```zsh
brew install deno
```

### Go

TBD

### Rust

TBD
